Source/Particles/Collision/BinaryCollision/DSMC/DSMC.cpp

```diff
@@ -0,0 +1,300 @@
+
+    // How much weight should be given to the produced particle - based on the
+    // weight of the collision partner which is not split
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> pair_reaction_weight(n_total_pairs);
+    amrex::ParticleReal* AMREX_RESTRICT p_pair_reaction_weight =
+                                                        pair_reaction_weight.dataPtr();
+
+    // Loop over cells
+    amrex::ParallelForRNG( n_cells,
+        [=] AMREX_GPU_DEVICE (int i_cell, amrex::RandomEngine const& engine) noexcept
+        {
+            // The particles from species1 that are in the cell `i_cell` are
+            // given by the `indices_1[cell_start_1:cell_stop_1]`
+            index_type const cell_start_1 = cell_offsets_1[i_cell];
+            index_type const cell_stop_1  = cell_offsets_1[i_cell+1];
+            // Same for species 2
+            index_type const cell_start_2 = cell_offsets_2[i_cell];
+            index_type const cell_stop_2  = cell_offsets_2[i_cell+1];
+            // Same but for the pairs
+            index_type const cell_start_pair = p_pair_offsets[i_cell];
+
+            // ux from species1 can be accessed like this:
+            // ux_1[ indices_1[i] ], where i is between
+            // cell_start_1 (inclusive) and cell_start_2 (exclusive)
+
+            // Do not collide if one species is missing in the cell
+            if ( cell_stop_1 - cell_start_1 < 1 ||
+                cell_stop_2 - cell_start_2 < 1 ) { return; }
+
+            // shuffle
+            ShuffleFisherYates(indices_1, cell_start_1, cell_stop_1, engine);
+            ShuffleFisherYates(indices_2, cell_start_2, cell_stop_2, engine);
+#if defined WARPX_DIM_RZ
+            int ri = (i_cell - i_cell%nz) / nz;
+            auto dV = MathConst::pi*(2.0_prt*ri+1.0_prt)*dr*dr*dz;
+#endif
+            // Call the function in order to perform collisions
+            // If there are product species, p_mask, p_pair_indices_1/2, and
+            // p_pair_reaction_weight are filled here
+            CollisionFilter(
+                cell_start_1, cell_stop_1, cell_start_2, cell_stop_2,
+                indices_1, indices_2,
+                ptd_1, ptd_2,
+                m1, m2, dt, dV,
+                cell_start_pair, p_mask, p_pair_indices_1, p_pair_indices_2,
+                p_pair_reaction_weight,
+                process_count, scattering_processes, engine );
+        }
+    );
+
+    const auto num_p_tile1 = ptile_1.numParticles();
+    const auto num_p_tile2 = ptile_2.numParticles();
+
+    // Create the new product particles and define their initial values
+    // num_added: how many particles of each product species have been created
+    const int num_added = splitScatteringParticles(
+        n_total_pairs,
+        ptile_1, ptile_2,
+        p_mask,
+        copy_species1, copy_species2,
+        m1, m2,
+        p_pair_indices_1, p_pair_indices_2,
+        p_pair_reaction_weight);
+
+    if (num_added > 0) {
```

Comment:
That is correct yes since we don't know which species had the lower weight during the collision. So we add new particles for both species but for one there will also be a particle removed (weight -> 0 and the ID set to negative), resulting in no net increase in the total number of particles for that species.